### ISSUE TITLE
Trying to `put` but `stat()` callback is returning undefined for both `err` and `stats`

### ISSUE THREAD

**Issue by rosschapman** [theophilusx/ssh2-sftp-client#274]
I receive an error when the code reaches https://github.com/theophilusx/ssh2-sftp-client/blob/master/src/index.js#L277 because the `stats` object is `undefined`. I don't have details about the remote server (yet) which is controlled by a third party, but perhaps you may have some instinct as to why the callback for `this.sftp.stat` would return `undefined` for both `err` and `stats` arguments. 

```
- Version of node? ==> v12.18.4
- Version of ssh2-sftp-client? ==> latest
- Platform client is on? ==> N/A
- Platform remote sftp server is on? ==> N/A
```

**Comment by theophilusx** [comment]
This is an error I've seen reported when the code is either not waiting
for the Promise to resolve or is using a 'then' chain where one of the
blocks is failing to return a promise.

Can you provide a minimal example of your code which is failing with
this error? I may be able to spot the issue quickly with an example.

**Comment by rosschapman** [comment]
I'm using `async/await` so now I wonder if values are returned as expected in your code:

```js
var buffer = Buffer.from(sample, "utf-8");
let client = new Client();

try {
  await client.connect({
    host: 'sample',
    port: 'sample',
    username: 'sample',
    password: 'sample',
    debug: msg => {
      if (msg.startsWith('CLIENT')) {
        shared_c.log.debug('DEBUG: ', msg)}
    }
  })
  let cwd = await client.cwd();
  shared_c.log.debug('Current dir: ', cwd);
  await client.put(buffer, `${cwd}/sample.dat`);
  // I've also tried ⤵
  // return await client.put(buffer, `${cwd}/test_ne_report_3.dat`);
} catch(e) {
  console.log(e, 'catch error');
} finally {
  await client.end();
}
```

Error log: 
```
04:18:22.838Z DEBUG binlog: DEBUG:  CLIENT[sftp]: stats <-  (index=dev_binlog, gitrev=00000000)
04:18:22.844Z ERROR binlog:  (index=dev_binlog, gitrev=00000000, console=error)
04:18:22.844Z ERROR binlog: (index=dev_binlog, gitrev=00000000, console=error)
  /node_modules/ssh2-streams/lib/sftp.js:389
                  cb();
                  ^
04:18:22.844Z ERROR binlog: (index=dev_binlog, gitrev=00000000, console=error)
  TypeError: Cannot read property 'mode' of undefined
      at /node_modules/ssh2-sftp-client/src/index.js:277:27
      at SFTPStream._transform (/node_modules/ssh2-streams/lib/sftp.js:389:17)
      at SFTPStream.Transform._read (_stream_transform.js:191:10)
      at SFTPStream._read (/node_modules/ssh2-streams/lib/sftp.js:183:15)
      at SFTPStream.Transform._write (_stream_transform.js:179:12)
      at doWrite (_stream_writable.js:403:12)
      at writeOrBuffer (_stream_writable.js:387:5)
      at SFTPStream.Writable.write (_stream_writable.js:318:11)
      at Channel.ondata (_stream_readable.js:716:22)
      at Channel.emit (events.js:315:20)
      at Channel.EventEmitter.emit (domain.js:483:12)
      at addChunk (_stream_readable.js:295:12)
      at readableAddChunk (_stream_readable.js:271:9)
      at Channel.Readable.push (_stream_readable.js:212:10)
      at SSH2Stream.<anonymous> (/node_modules/ssh2/lib/Channel.js:167:15)
      at SSH2Stream.emit (events.js:315:20)
      at SSH2Stream.EventEmitter.emit (domain.js:483:12)
      at parsePacket (/node_modules/ssh2-streams/lib/ssh.js:3191:10)
      at SSH2Stream._transform (/node_modules/ssh2-streams/lib/ssh.js:701:13)
      at SSH2Stream.Transform._read (_stream_transform.js:191:10)
      at SSH2Stream._read (/node_modules/ssh2-streams/lib/ssh.js:253:15)
      at SSH2Stream.Transform._write (_stream_transform.js:179:12)
      at doWrite (_stream_writable.js:403:12)
      at writeOrBuffer (_stream_writable.js:387:5)
      at SSH2Stream.Writable.write (_stream_writable.js:318:11)
      at Socket.ondata (_stream_readable.js:716:22)
      at Socket.emit (events.js:315:20)
      at Socket.EventEmitter.emit (domain.js:483:12)
```

**Comment by theophilusx** [comment]
The below line is very odd - never seen that in debug before.

> 04:18:22.844Z ERROR binlog: (index=dev_binlog, gitrev=00000000, console=error)
>   /node_modules/ssh2-streams/lib/sftp.js:389
>                   cb();
>                   ^

What happens if

- You use openSSH sftp client to upload a sample file to the same
  location?
- Use openSSH sftp client to list contents of cwd?
- Use ssh2-sftp-client to upload using a filename rather than a buffer?
- Use ssh2-sftp-client to call stat() on the value returned by cwd()?

Also, if you look at the debug output, right at the very beginning, it
normally provides some details about the remote server. That info might
be helpful in working out what the issue might be.

**Comment by rosschapman** [comment]
Suggestion #1 and #3 works. Suggestion #4 does not work. I didn't get around to testing #2. Because of time pressure we went ahead and wrote a small wrapper around `ssh2` directly to satisfy our downscoped needs so I won't be able to help debug this further. I wasn't able to see any debug info from the server in the logs but it's possible the logging in our development infrastructure is customized in a way that it might filter out certain logs. Thanks for the speedy replies. It's possible this is a bit of a one-off fluke related to some server config on the third-party side.

**Comment by rosschapman** [comment]
Ok I had a few minutes and went back and opened a connection and at the `sftp>` prompt played around with trying to stat the root directory where I'd been trying to upload files. That server does not support `stat` but supports `df` and trying that I received this helpful error message: `Server does not support statvfs@openssh.com extension`. I suspect this is relevant and there may be guards on the server to prevent reading the filesystem stats?

**Comment by theophilusx** [comment]
Thanks for doing that, really appreciated.

This was my suspicion.

A couple of versions back, I added a lot of additional error checking to
try and provide more informative error messages to the client. Many of
the error messages from ssh2/ssh2-streams are not particularly
informative and led to user confusion (e.g. error messages such as
'Failure').

To do this, I added code to do things like verify remote files exist or
don't exist or remote target is a file or a directory etc. This allowed
me to provide better error messages that helped users and based on the
drop in basic issues, this seems to have worked.

Unfortunately, this has had two undesirable side effects. Firstly, not
all sftp servers are equal and some don't support stat, realPath and
listing. While these servers are in the minority, they do exist. The
second problem is that this also has a bit of a hit on performance.
The hit on performance is particularly noticeable on slow connections.

So I made the decision to remove all this additional error checking in
the next version. At the same time, the ssh2 module is getting a
complete re-write (so I'm hoping I might be able to convince the
maintainer to add/improve error messages where necessary). The
ssh2-streams library is being deprecated and its functionality moved
into the main ssh2 module.

So the next version will work with sftp servers like the one you are
using. Unfortunately, I have to wait for some of the additional
functionality (like streams) to be added to ssh2. The ssh2 author says
to use ssh2-streams if you still need stream support, but one of the
reasons for the ssh2 re-write is that ssh2-streams does not work
correctly with node v14+. It is likely I may need to re-think some of
the design and API to work with the new ssh2 and I will need to wait
until the new ssh2 reaches production readiness before a new version can
be released. If this takes too long, I might have to release an
intermediate versions which has removed the additional error checking,
but still uses the existing ssh2/ssh2-streams modules (which means no
node v14+ support).

I've started working on the new versions as the branch version-6 in the
git repository. The new version should work with sftp servers like the
one your using, but it won't be production ready for quite some time.
